feat(tutorials): add C++ casino number guessing game tutorial

biru-codeastromer opened the pull request:
#### Fixes #5881 

### Description
This PR adds a new tutorial for building a C++ application using Jenkins. The tutorial demonstrates how to set up a **Casino Number Guessing Game** project, configure a Jenkins pipeline, and automate the build, test, and delivery process.

### Changes Included
1. **Tutorial Documentation**:
   - Step-by-step instructions for setting up the project.
   - Detailed explanation of the `Jenkinsfile` with Build, Test, and Deliver stages.
   - Screenshots for each stage of the pipeline.

2. **Repository**:
   - Added a new repository (`casino-number-guessing-game`) with two branches:
     - `main`: Skeleton code for users to implement.
     - `solution`: Complete solution with the C++ code and `Jenkinsfile`.

3. **Docker Setup**:
   - Bruno Sir included a Docker profile for C++ in the `quickstart-tutorials` repository.
      Thankful to him

### Testing
- The tutorial has been tested end-to-end on a local Jenkins setup.
- The pipeline successfully builds, tests, and delivers the C++ application.
- 

https://github.com/user-attachments/assets/d03e8e29-3fdf-40cd-adb3-b7a8550d80f6


### Next Steps
- Once approved, the repository will be migrated to the `jenkins-docs` .


Looking forward to your feedback!

biru-codeastromer commented:
Fixing the ci checks soon !

biru-codeastromer commented:
Done ! corrected the ci checks .

krisstern commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+    }
+}
+----
+Defines a new stage called `Deliver` that appears on the Jenkins UI.
```
```suggestion
The above defines a new stage called `Deliver` that appears on the Jenkins UI.
```

biru-codeastromer commented:
Thanks for all your suggestions 👍.
I have implemented them all .

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+. Select *Definition* and then choose the *Pipeline script from SCM* option. This option instructs Jenkins to obtain your Pipeline from the source control management (SCM), which is your forked Git repository.
+. Choose *Git* from the options in *SCM*.
+. Enter the URL of your repository in *Repositories/Repository URL*. This URL can be found when selecting the green *Code* button in the main page of your GitHub repo.
+. Set the script path to `Jenkinsfile`.
```
```suggestion
. In _Branches to build_, enter `*/main`
. In _Script Path_, set the script path to `Jenkinsfile`.
```

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+
+This is the foundation of "Pipeline-as-Code", which treats the continuous delivery pipeline as a part of the application to be versioned and reviewed like any other code. Read more about Pipeline and what a Jenkinsfile is in the link:/doc/book/pipeline[Pipeline] and link:/doc/book/pipeline/jenkinsfile[Using a Jenkinsfile] sections of the User Handbook.
+
+First, create an initial Pipeline to download a C++ Docker image and run it as a Docker container (which will build your C++ application). Also add a "Build" stage to the Pipeline that begins orchestrating this whole process.
```
```suggestion
First, create an initial Pipeline to build your C++ application. Add a "Build" stage to the Pipeline that begins orchestrating this whole process.
```

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+
+First, create an initial Pipeline to download a C++ Docker image and run it as a Docker container (which will build your C++ application). Also add a "Build" stage to the Pipeline that begins orchestrating this whole process.
+
+. Using your favorite text editor or IDE, create and save a new text file with the name `Jenkinsfile` at the root of your local `casino-number-guessing-game` Git repository.
```
```suggestion
. As you forked our sample repo, you already have an empty `Jenkinsfile`.
```

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+. Initially queues the project to be run on the agent.
+. Runs the `Build` stage defined in the `Jenkinsfile` on the agent.
+
+During this time, CMake downloads and configures the necessary build files for your C++ application. You can now select *#1* to see the details of the build. You will then see how much time the build took waiting in the queue, and how much time it took to run.
```
```suggestion
During this time, CMake configures the necessary build files for your C++ application, builds and links your application.
You can now select *#1* to see the details of the build. You will then see how much time the build took waiting in the queue, and how much time it took to run.
```

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-archiveartifacts-code-archive-the-artifacts[`archiveArtifacts`] step archives the `.tar.gz` file as a build artifact.
+
+. Save your edited `Jenkinsfile` and commit it to your local `casino-number-guessing-game` Git repository. Within the `casino-number-guessing-game` directory, run the commands: +
+  `git add .` +
```
```suggestion
  `git add Jenkinsfile` +
```

gounthar commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,314 @@
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script[`sh`] step executes the `cmake` command to build your C++ application.
+
+. Save your edited `Jenkinsfile` and commit it to your local `casino-number-guessing-game` Git repository. Within the `casino-number-guessing-game` directory, run the commands: +
+  `git add .` +
```
```suggestion
  `git add Jenkinsfile` +
```

biru-codeastromer commented:
> Very good job! 🚀 I tried your tutorial from the beginning to the end, and I changed a few things along the way.
> 
> Thanks a lot!



Thanks for your appreciation !  Glad that you liked the tutorial .
I have implemented your suggestions and welcome from my side .

gounthar reviewed (approved):
Great work, thanks for your contribution. 👍

krisstern commented:
Just one comment though, should we put the C++ tutorial on the top of the list? Or should we put it more towards the bottom of the list at `/doc/tutorials/`?

biru-codeastromer commented:
> Great work, thanks for your contribution. 👍

You are welcome  🤗

krisstern commented:
@kmartens27 do you have any preference regarding the ordering of the tutorials? Somehow I don't think it should be put up top due to the comparative popularity of the programming languages. What do you think?

krisstern commented:
I am thinking of something like the TIOBE Index when I posed the question earlier: https://www.tiobe.com/tiobe-index/
Maybe it is a good time to reorder the tutorials too based on the trending popularities?

biru-codeastromer commented:
> I was thinking of something like the TIOBE Index when I posed the question earlier: https://www.tiobe.com/tiobe-index/ Maybe it is a good time to reorder the tutorials too based on the trending popularities?



I loved this idea . It will be very good practice to follow that . 
Thanks for sharing that useful link here .

kmartens27 commented:
I also like the suggestion from @krisstern about reordering the tutorials based on the popularity/TIOBE index.

biru-codeastromer commented:
Hi @kmartens27,

Thank you for going through the tutorial and sharing your feedback! 😊

I appreciate your point about the output. The current tutorial mainly focuses on setting up the Jenkins pipeline for a C++ project, but I agree that having a visible output (like the generated number) would make it more complete.

I can update the tutorial to include a simple step where the program prints the generated number in the build logs. This will ensure that users can see an immediate result without diving into the artifacts. Would that work?


Looking forward to your additional suggestions too!

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+[.boxshadow]
+image:tutorials/cpp/deliver-overview.png[alt="Deliver stage",width=100%]
+
+. Select *#3* or on the number representing your last build on the left, under *Build History*. 
```
```suggestion
Select *#3*, or on the number representing your last build on the left, under *Build History*. 
```
Remove numbered step as it is not part of the step-by-step instructions, add comma to include text better.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+
+. In Jenkins, sign in if necessary, go back to the *Dashboard* and then navigate to *casino-number-guessing-game*. Alternatively, you can go directly to *casino-number-guessing-game* depending on where you're starting from.
+. Select *Build Now* on the left. You should see after a while a new column *Deliver* appear in the *Stage View*.
+
```
```suggestion
+
```
Adding a `+` will keep everything in line with the previous steps instead of starting new steps each time.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+image:tutorials/cpp/deliver-overview.png[alt="Deliver stage",width=100%]
+
+. Select *#3* or on the number representing your last build on the left, under *Build History*. 
+
```
```suggestion
+
```
Same as above, keeps everything in line with following steps.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+image:tutorials/cpp/deliver-overview-status.png[alt="Deliver stage",width=100%]
+
+. Select *Pipeline Overview* to see the stages of the Pipeline.
+
```
```suggestion
+
```
Same as above, keep everything in line with following steps.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+image:tutorials/cpp/pipeline-overview-deliver-stage.png[alt="Pipeline overview with Deliver stage",width=100%]
+
+. Select the *Deliver* stage. You will then see a green part displaying *tar -czf casino_game.tar.gz build/casino_game*, which represents the successful execution of the `tar` command.
+
```
```suggestion
+
```
Same as above, keep everything in line with previous steps.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+image:tutorials/cpp/pipeline-console-deliver-stage.png[alt="Deliver stage output",width=100%]
+
+. You can see the full stage view by clicking on the "Full Stage View" in the left menu.
+
```
```suggestion
+
```
Same as above, keep everything in line with previous steps.

kmartens27 commented:
In regards to your previous comment, I think the number generator would be ideal, and the intro text can be adjusted to frame that. Since there isn't any instruction or step that tells a reader to investigate the artifacts, this would help prevent any confusion or issues. I know the goal is to showcase how C++ can be used with Jenkins, but we should be cautious about what expectations we're setting vs. delivering. Some people may not even care about that part of it, but we can account for any noise by ensuring that the text is accurate.

krisstern commented:
Regarding the example code for C++, I just think a "Hello, world!" one in general is too simplistic and does not demonstrate much by way of the user's interaction with the application itself. That's my rationale when suggesting the idea currently used in this PR.

biru-codeastromer commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+. Copy the following Declarative Pipeline code and paste it into your empty `Jenkinsfile`:
++
+[source,groovy,subs="attributes+"]
+----
+pipeline {
+    agent any
+    stages {
+        stage('Build') {
+            steps {
+                sh 'rm -rf build'
+                sh 'cmake -B build -S .'
+                sh 'cmake --build build'
+            }
+        }
+    }
+}
+----
+The above defines a link:/doc/book/pipeline/syntax/#stage[`stage`] (directive) called `Build` that appears on the Jenkins UI.
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script[`sh`] step executes the `cmake` command to build your C++ application.
```
Hi @kmartens27  , I guess those symbols somehow disturbs the ci checks as in my initial commit it did create problems . So I fear of implementing this suggestion . What you believe??

biru-codeastromer commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+                sh 'cmake --build build'
+            }
+        }
+        stage('Test') {
+            steps {
+                sh './build/casino_game'
+                sh './build/test_game'
+            }
+        }
+    }
+}
+----
+The above defines a link:/doc/book/pipeline/syntax/#stage[`stage`] (directive) called `Test` that appears on the Jenkins UI.
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script[`sh`] step runs the game and unit tests to ensure the application works as expected.
```
same point here I want to put . Those symbols somehow disturbs the ci checks as in my initial commit it did create problems . So I fear of implementing this suggestion .

biru-codeastromer replied:
Thanks so much for the explanation about this .Much appreciated . I misinterpreted it . Sorry from my side . Now I got to know where the issue lied . 

Let me try to implement them now .

biru-codeastromer commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,316 @@
+                sh './build/test_game'
+            }
+        }
+        stage('Deliver') {
+            steps {
+                sh 'tar -czf casino_game.tar.gz build/casino_game'
+                archiveArtifacts artifacts: 'casino_game.tar.gz', fingerprint: true
+            }
+        }
+    }
+}
+----
+The above defines a new stage called `Deliver` that appears on the Jenkins UI.
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script[`sh`] step packages the executable into a `.tar.gz` file.
+This link:/doc/pipeline/steps/workflow-durable-task-step/#code-archiveartifacts-code-archive-the-artifacts[`archiveArtifacts`] step archives the `.tar.gz` file as a build artifact.
```
same point here I want to put too again like those two previous . Those symbols somehow disturbs the ci checks as in my initial commit it did create problems . So I fear of implementing this suggestion .

biru-codeastromer commented:
Hello @kmartens27 I have implemented all your suggestions (except 3 of them) as I found them useful and also more about Jenkins documentation format. Thanks for that !

One thing I want to point out those 3 suggestions that I didn't executed , for which ci checks failed , I fear to implement those .
Let me know what you think about it.
I faced that issue so I  informed you about that .

biru-codeastromer commented:
Hi @kmartens27  @krisstern @gounthar ,

I wanted to check in with you three on an important question.

There are two possible ways to address this:
1. Updating the intro text to guide users toward checking artifacts for output.
2. Implementing a number display feature in the build logs for immediate feedback (as Kevin suggested).

Since expectation management is key here, I’d love to hear your thoughts on which approach would be more suitable. Should we modify the instructions to align expectations, or would making the number generation explicitly visible in the logs be a better approach?

I personally feel that either option could work well for users, but I might be overlooking something. Would love to hear your insights!

biru-codeastromer commented:
> I was thinking of something like the TIOBE Index when I posed the question earlier: https://www.tiobe.com/tiobe-index/ Maybe it is a good time to reorder the tutorials too based on the trending popularities?

Also should I rearrange the order in this manner now and push the change ?

kmartens27 commented:
@biru-codeastromer If the artifacts would contain the generated number, it would be good to include that information in the tutorial. I don't think incorporating a number display feature would be necessary. It could be just a short line or two explaining where the result would be available/how to find it.

The reason I brought this up was that I thought I had done something wrong since I didn't see any results/numbers, and fear that this could be something others experience as well. If we inform readers that the number would be generated and available in the artifacts then that would resolve the issue.

biru-codeastromer commented:
@kmartens27 Hi, there is a happy news now !

I have modify the logic in the solution branch of the repo to have actually showed a number  and now in the test stage output we can actually see the whole game being played which was not present before ☺️

biru-codeastromer commented:
<img width="1028" alt="Screenshot 2025-02-27 at 12 57 39 AM" src="https://github.com/user-attachments/assets/92a2d111-7df3-48ca-87ce-3fb473a21eb1" />


This is the update let me how it looks now ?? 🎉

biru-codeastromer commented:
> @biru-codeastromer If the artifacts would contain the generated number, it would be good to include that information in the tutorial. I don't think incorporating a number display feature would be necessary. It could be just a short line or two explaining where the result would be available/how to find it.
> 
> The reason I brought this up was that I thought I had done something wrong since I didn't see any results/numbers, and fear that this could be something others experience as well. If we inform readers that the number would be generated and available in the artifacts then that would resolve the issue.




I implemented your previous suggestion to actually see how will it looks and rather it worked . I have changed the code to have the **number generation** present .

Does it now aligns with your previous suggestion ? Let me know about that .

biru-codeastromer commented:
> It looks like everything processed through correctly this time! 🎉 The only thing I wasn't sure of with the callouts was the actual step that they should be in line with. I assumed that <img alt="Screenshot 2025-02-26 at 2 18 08 PM" width="473" src="https://private-user-images.githubusercontent.com/99040580/417294565-2a320d11-f455-4cf4-a5fb-bb6fa437fad5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1OTc4NDAsIm5iZiI6MTc0MDU5NzU0MCwicGF0aCI6Ii85OTA0MDU4MC80MTcyOTQ1NjUtMmEzMjBkMTEtZjQ1NS00Y2Y0LWE1ZmItYmI2ZmE0MzdmYWQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE5MTkwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzMzY0ZjUyNmE0NTM3OTAwNWI4MjNhMWVmYTUwZTM0YTc5ZDAxOGM3YThhMzdmMjcxZTg3ZjM5ZTg4YzQ0NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xF5TILUCE3sB2gOdekarAUTK6YyA5-3WYdd9h2900iA"> and <img alt="Screenshot 2025-02-26 at 2 18 00 PM" width="480" src="https://private-user-images.githubusercontent.com/99040580/417294657-cfcc8ede-8136-4926-88e4-7bd81c434690.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1OTc4NDAsIm5iZiI6MTc0MDU5NzU0MCwicGF0aCI6Ii85OTA0MDU4MC80MTcyOTQ2NTctY2ZjYzhlZGUtODEzNi00OTI2LTg4ZTQtN2JkODFjNDM0NjkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE5MTkwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1ZmZhYjgxZWM5YmU4ZTMyMjI1YTE0YmYyZWFhNDAwNGFkNDU5NTA5OTEwZDU5YTY5NjlhMDI1ZTIwYjkxZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5tAukDLZNy7TQc0z9Ky1WYwq6DX2-ZgLuVWBk2vR4fQ">
> 
> were the correct lines to callout, but if those are not accurate, please update them to the correct lines.



Sure let me check it and confirm .

biru-codeastromer commented:
Previously : 

<img width="965" alt="Screenshot 2025-02-27 at 1 09 15 AM" src="https://github.com/user-attachments/assets/c79edf3e-1382-4afb-bf7e-1571138e1a20" />


Now after changes: 

<img width="1028" alt="Screenshot 2025-02-27 at 12 57 39 AM" src="https://github.com/user-attachments/assets/dd857c01-ad8a-46b8-98fa-6f4caa852991" />

kmartens27 commented:
Thanks very much @biru-codeastromer, that looks great and now that there are "results" it looks much more complete. I appreciate you taking the time to create and test all of that!

biru-codeastromer commented:
Welcome @kmartens27 ,

I have added a view results section for users to know how they can see the results.
Let me know how you think about the latest changes in the documentation after review . Thanks !

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,377 @@
+[.boxshadow]
+image:tutorials/cpp/pipeline-output-test-stage.png[alt="Test stage output",width=100%]
+
+==== Viewing the Results
```
```suggestion
==== Viewing the results
```
Suggestion to use sentence casing (only beginning of header and any propper nouns).

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,377 @@
+
+During the **Test** stage, the `casino_game` executable will run and display output similar to the following in the Jenkins console:
+
+[source,bash]
+----
+Welcome to the Casino Number Guessing Game!
+Running in non-interactive mode. Using default values.
+Generated random number: 8
+Generated random number: 4
+Generated number: 4
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Wrong guess! Try again.
+Congratulations! You guessed the number!
+You won $5!
+Total winnings: $5
+----
+
+This output confirms that the application is working as expected. Here’s what each part of the output means:
+
+- **`Generated random number`**: This line shows the random number generated by the game. In non-interactive mode, the game generates multiple random numbers to simulate guesses until it finds the correct one.
+- **`Generated number`**: This is the secret number the game is trying to guess.
+- **`Wrong guess! Try again.`**: This indicates that the game made an incorrect guess. In non-interactive mode, the game will continue guessing until it finds the correct number.
+- **`Congratulations! You guessed the number!`**: This indicates that the game successfully guessed the correct number.
+- **`You won $5!`**: This shows the prize awarded based on the number of attempts and the difficulty level.
+- **`Total winnings: $5`**: This displays the total winnings accumulated during the game.
```
```suggestion
[source,bash]
----
Welcome to the Casino Number Guessing Game!
Running in non-interactive mode. Using default values.
Generated random number: 8 // <1>
Generated random number: 4
Generated number: 4 // <2>
Wrong guess! Try again. // <3>
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Wrong guess! Try again.
Congratulations! You guessed the number! // <4>
You won $5! // <5>
Total winnings: $5 // <6>
----

This output confirms that the application is working as expected. Here’s what each part of the output means:

<1> **`Generated random number`**: This line shows the random number generated by the game. In non-interactive mode, the game generates multiple random numbers to simulate guesses until it finds the correct one.
<2> **`Generated number`**: This is the secret number the game is trying to guess.
<3> **`Wrong guess! Try again.`**: This indicates that the game made an incorrect guess. In non-interactive mode, the game continues guessing until it finds the correct number.
<4> **`Congratulations! You guessed the number!`**: This indicates that the game successfully guessed the correct number.
<5> **`You won $5!`**: This shows the prize awarded based on the number of attempts and the difficulty level.
<6> **`Total winnings: $5`**: This displays the total winnings accumulated during the game.
```
Suggestion to add callouts to this block, tested in my local clone and the site does build with the callouts as expected. Small text suggestion as well.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,377 @@
+- **`You won $5!`**: This shows the prize awarded based on the number of attempts and the difficulty level.
+- **`Total winnings: $5`**: This displays the total winnings accumulated during the game.
+
+The `test_game` executable will also run during the **Test** stage and display output similar to the following:
```
```suggestion
The `test_game` executable also runs during the **Test** stage and display output similar to the following:
```
Suggestion to use active voice.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,377 @@
+
+This output confirms that the unit tests for the game logic have passed successfully.
+
+For reference , here is how the output will look like : 
```
```suggestion
For reference, this is an example of how the output renders: 
```
Suggestion to use present voice.

kmartens27 commented on content/doc/tutorials/build-a-cpp-app-with-jenkins.adoc:
```diff
@@ -0,0 +1,377 @@
+
+==== Viewing the Results
+
+During the **Test** stage, the `casino_game` executable will run and display output similar to the following in the Jenkins console:
```
```suggestion
During the **Test** stage, the `casino_game` executable runs and displays output similar to the following in the Jenkins console:
```
Suggestion to use present/active voice.

kmartens27 reviewed (approved):
Thanks very much for all the work and updates on this @biru-codeastromer, it looks great!

biru-codeastromer commented:
Welcome @kmartens27  🤗 

Sorry for that future tense usage .I always forget that , I will try to keep in mind moving forward .
Thanks !

biru-codeastromer commented:
about the order arrangement mentioned from https://www.tiobe.com/tiobe-index/ 

how about this order proposal ? 

```
* link:build-a-python-app-with-pyinstaller[Build a Python app with PyInstaller]
* link:build-a-cpp-app-with-jenkins[Build a C++ app with Jenkins]
* link:build-a-java-app-with-maven[Build a Java app with Maven]
* link:build-a-node-js-and-react-app-with-npm[Build a Node.js and React app with npm]
* link:using-jenkinsfile-runner-github-action-to-build-jenkins-pipeline[Build a Jenkins pipeline by using Jenkinsfile Runner GitHub Actions]
* link:build-a-labview-app[Build a LabVIEW app]
```

kmartens27 commented:
No worries on the future tense! As far as the order of tutorials, that looks like a solid order to me (especially based on the TIOBE stats/preferences.

biru-codeastromer commented:
> No worries on the future tense! As far as the order of tutorials, that looks like a solid order to me (especially based on the TIOBE stats/preferences.



Thanks !

krisstern commented on content/doc/tutorials/index.adoc:
```diff
@@ -42,6 +42,7 @@ concepts based on specific technology stacks.
 Choose the tutorial that's relevant to your technology stack or one that you're
 most familiar with:
 
+* link:build-a-cpp-app-with-jenkins[Build a C++ app with Jenkins]
 * link:build-a-java-app-with-maven[Build a Java app with Maven]
 * link:build-a-node-js-and-react-app-with-npm[Build a Node.js and React app with npm]
 * link:build-a-python-app-with-pyinstaller[Build a Python app with PyInstaller]
```
Please update the ordering according to the TIOBE index at https://www.tiobe.com/tiobe-index/ as discussed.

krisstern replied:
So ideally Python > C++ > Java > Node.js

krisstern replied:
We added the Jenkinsfile tutorial I think two years ago in 2022 following the successful completion of the relevant GSoC project. But due to the lack of maintenance of the Jenkinsfile Runner repo it may not work anymore. So better put it last.

biru-codeastromer replied:
Got it

biru-codeastromer replied:
Thanks for the order confirmation I will do it now .

krisstern replied:
Sounds good!

biru-codeastromer commented:
Done ! I have made the change .

A small question about the last tutorial  on : 

< But due to the lack of maintenance of the Jenkinsfile Runner repo it may not work anymore.

If it won't work , I want to know why we are keeping that tutorial still in place . Shouldn't it be removed or maybe archive ?

krisstern commented:
> If it won't work, I want to know why we are keeping that tutorial still in place. Shouldn't it be removed or maybe archive?

It is because we won't have any other records of it in the docs otherwise, which is worst.

biru-codeastromer commented:
I see the case now . Thanks for your clarification 😊

biru-codeastromer commented:
> Thanks for the PR @biru-codeastromer!

Welcome 🤗

kmartens27 commented:
Hi @biru-codeastromer just wanted to share the compressed version of the images here for usage. I know there is the image compress action, but to be sure I wanted to see if you would be able to utilize these first.

[c++ first 10.zip](https://github.com/user-attachments/files/18996349/c%2B%2B.first.10.zip)
[c++ last 6.zip](https://github.com/user-attachments/files/18996359/c%2B%2B.last.6.zip)

biru-codeastromer commented:
> Hi @biru-codeastromer just wanted to share the compressed version of the images here for usage. I know there is the image compress action, but to be sure I wanted to see if you would be able to utilize these first.
> 
> [c++ first 10.zip](https://github.com/user-attachments/files/18996349/c%2B%2B.first.10.zip) [c++ last 6.zip](https://github.com/user-attachments/files/18996359/c%2B%2B.last.6.zip)



Thanks a lot ! I will implement them .

biru-codeastromer commented:
@kmartens27 Thanks !

 I have implemented the suggestion by adding those images .

biru-codeastromer commented:
You are welcome 😊 @kmartens27 . And thanks to you , @krisstern  and @gounthar for all your valuable feedbacks on this PR . Much appreciated !

gounthar commented:
That was a great collaborative work, thanks a lot to all participants. 🤗

gounthar commented:
Thanks @zbynek!
Yes, the repo should be transferred to the [jenkins-docs organization](https://github.com/jenkins-docs) once it's settled.
@biru-codeastromer : this request should be done via the creation of an [issue in the helpdesk](https://github.com/jenkins-infra/helpdesk/issues).

biru-codeastromer commented:
Hello @zbynek Thanks for your follow-up .


Regards to the unwanted files , I’ll clean up the build/ folder and remove the unnecessary binaries. Thanks for pointing that out!

biru-codeastromer commented:
> Thanks, @zbynek! Yes, the repo should be transferred to the [jenkins-docs organization](https://github.com/jenkins-docs) once it's settled. @biru-codeastromer : this request should be done via the creation of an [issue in the helpdesk](https://github.com/jenkins-infra/helpdesk/issues).



Sure I will create an issue on that once I have cleaned up my repo and informed you there . Thanks !

biru-codeastromer commented:
Hi @zbynek , regarding the 10% probability case , I have made a refined logic to use : 

```
#include <iostream>
#include "../src/Game.h"
#include <cassert>

void testGame() {
    Game game(1); // Easy difficulty

    int secretNumber = game.generateRandomNumber();

    assert(secretNumber >= 1 && secretNumber <= 10);

    int wrongGuess = (secretNumber == 1) ? 2 : 1;  
    assert(game.checkGuess(wrongGuess) == false);

    assert(game.checkGuess(secretNumber) == true);
}

int main() {
    testGame();
    std::cout << "All tests passed!" << std::endl;
    return 0;
}
```


Let me know how you think about this logic?? Also we can maybe use a seeded random number generator approach .

What you all think would be best in regards to this let me know about that . Thanks !

kmartens27 commented:
That's a very good point @zbynek thanks for calling that out. Thanks @biru-codeastromer for being open to sorting that out once the repo is ready to pass over.

biru-codeastromer commented:
The  seeded random number generator approach I mentioned is thinking of maybe using this approach : 

```
void testGame() {
    srand(50); 

    Game game(1); // Easy difficulty

    int secretNumber = game.generateRandomNumber();

    assert(secretNumber >= 1 && secretNumber <= 10);

    int wrongGuess = (secretNumber == 1) ? 2 : 1;  
    assert(game.checkGuess(wrongGuess) == false);

    assert(game.checkGuess(secretNumber) == true);

    assert(game.checkGuess(1) == (secretNumber == 1));
    assert(game.checkGuess(10) == (secretNumber == 10));
}
```

It makes the test fully deterministic because the random number generator is seeded.

biru-codeastromer commented:
If you believe it is much better approach in the test , then I will implement this change in the repo  .

zbynek commented:
The first approach looks better, maybe even simpler with `int wrongGuess = secretNumber + 1`.

biru-codeastromer commented:
Got it !
Let me implement it and send you the file for you to review how it looks .

biru-codeastromer commented:
Came here to if everything is good in the repo now and I get approved about that , I will raise the issue in infra repo to include that repo in jenkins docs repo . Thanks !

krisstern commented:
@biru-codeastromer I noticed for the [code snippet](https://github.com/biru-codeastromer/casino-number-guessing-game/blob/c02c482f9d31a4011340f493b56b529b9aa7de55/src/main.cpp#L18-L22) in your "main.cpp" file where we have
```c++
  if (isNonInteractive) {
      // TODO: Use default values for non-interactive mode
  } else {
      // TODO: Prompt for user input in interactive mode
  }
```
looks like it is yet to be implemented. Is that intentional?

krisstern commented:
In fact the game looks incomplete in other places as well. I thought we are not asking the user for their input and we are supposed to be providing a well worked out solution for them to use in a "plug-and-play" like manner. Why is this not the case?

krisstern commented:
I mean I can see you have a `solution` in addition to a `main` branch. Why is that the case?

biru-codeastromer commented:
> @biru-codeastromer I noticed for the [code snippet](https://github.com/biru-codeastromer/casino-number-guessing-game/blob/c02c482f9d31a4011340f493b56b529b9aa7de55/src/main.cpp#L18-L22) in your "main.cpp" file where we have
> 
> ```c++
>   if (isNonInteractive) {
>       // TODO: Use default values for non-interactive mode
>   } else {
>       // TODO: Prompt for user input in interactive mode
>   }
> ```
> 
> looks like it is yet to be implemented. Is that intentional?



yes , it is intentional .

krisstern commented:
> I mean I can see you have a solution branch in addition to a main branch. Why is that the case?

Can we have just one branch?

biru-codeastromer commented:
> I mean I can see you have a `solution` branch in addition to a `main` branch. Why is that the case?



Hi @krisstern , because I was told to make two branch one with a skeleton and one with the solution to it . That's why in the main branch you are seeing places as "TODO "

krisstern commented:
> > I mean I can see you have a `solution` branch in addition to a `main` branch. Why is that the case?
> 
> Hi @krisstern , because I was told to make two branch one with a skeleton and one with the solution to it . That's why in the main branch you are seeing places as "TODO "

Who suggested it and why?

biru-codeastromer commented:
> > I mean I can see you have a solution branch in addition to a main branch. Why is that the case?
> 
> Can we have just one branch?



I made it as it was told to me in this comment : https://github.com/jenkins-infra/jenkins.io/issues/5881#issuecomment-2674537507

krisstern commented:
Also like for the [code block](https://github.com/biru-codeastromer/casino-number-guessing-game/blob/3e72106c4d7c40be80242414dc90290bd8e5c208/src/Game.cpp#L28C1-L37C2) in L28-L37 of "src/Game.cpp":

```c++
int Game::calculatePrize(int attempts) {
    int basePrize;
    switch (difficultyLevel) {
        case 1: basePrize = 100; break; // Easy
        case 2: basePrize = 500; break; // Medium
        case 3: basePrize = 1000; break; // Hard
        default: basePrize = 100;
    }
    return basePrize / attempts;
}
```

Is it intentional to force an `int` instead of a `float` and why?

krisstern commented:
> Did I misunderstood it or so ? If so I can just keep the main branch intact in that . Please see and have me confirm . Thanks !

I think the intention was and still is not to have a C++ puzzle for the user to solve, but to provide some ready-to-use code so people can experiment with Jenkins readily in a tutorial. So I do not see why we need two separate branches.

biru-codeastromer commented:
> I think the intention was and still is not to have a C++ puzzle for the user to solve, but to provide some ready-to-use code so people can experiment with Jenkins readily in a tutorial. So I do not see why we need two separate branches.



Got it . I think Bruno Sir might have suggested as to create a environment where users will try to make it and take help from solution branch to do and then use it with Jenkins . But I think as you said now , we do can remove the solution branch and just keep the main branch as the ultimate one ?? what you think ?? should i move ahead in doing that ? Let me know

biru-codeastromer commented:
Also one point I want to add and ask about that  having two branches is a common practice in tutorials to encourage hands-on learning , so let me know how you have views on this ?

biru-codeastromer commented:
> Also like for the [code block](https://github.com/biru-codeastromer/casino-number-guessing-game/blob/3e72106c4d7c40be80242414dc90290bd8e5c208/src/Game.cpp#L28C1-L37C2) in L28-L37 of "src/Game.cpp":
> 
> ```c++
> int Game::calculatePrize(int attempts) {
>     int basePrize;
>     switch (difficultyLevel) {
>         case 1: basePrize = 100; break; // Easy
>         case 2: basePrize = 500; break; // Medium
>         case 3: basePrize = 1000; break; // Hard
>         default: basePrize = 100;
>     }
>     return basePrize / attempts;
> }
> ```
> 
> Is it intentional to force an `int` instead of to have a `float` and why?




Thanks ! not intentional . Thanks for the point out I will change it to float .

krisstern commented:
@gounthar Did you suggest to have two branches for the code repo, with one for the skeleton and the other for the solution?If so, what is the motivation for doing so? I do not see why we need to ask the user to solution the problem as a challenge themselves, as that seems out of place. That is not what the tutorial is for to me.

biru-codeastromer commented:
Should I move in changing to float ? I used int to have a whole number  as value . If precision is the criteria I will change it .

krisstern commented:
> Should I move in changing to float ? I used int to have a whole number as value . If precision is the criteria I will change it .

Maybe using 2 decimal places for the float would make more sense.

biru-codeastromer commented:
> Also, why are the C++ file names capitalized?:
> 
> > For C++ files, prefer naming the file the same as the (main) class it contains. Currently all file names are all-lowercase, even though class names contain capital letters.
> 
> See: https://manual.gromacs.org/documentation/2019/dev-manual/naming.html#:~:text=For%20C%2B%2B%20files%2C%20prefer%20naming,class%20names%20contain%20capital%20letters.




Hi , I think it is told to use the lowercase/uppercase depending on how the class is name . I used "Game" , "Player" as classes so accordingly I named it as "Game.cpp" or "Player.cpp" and likewise for main i use lowercase as it was how the name in the codebase .

I think in that guide too , it is mentioned like that : 

> > For C++ files, prefer naming the file the same as the (main) class it contains.

biru-codeastromer commented:
> Maybe using 2 decimal places for the float would make more sense.

Got it , I will take this as one of the task for that repo . Thanks 😊

krisstern commented:
> Hi , I think it is told to use the lowercase/uppercase depending on how the class is name . I used "Game" , "Player" as classes so accordingly I named it as "Game.cpp" or "Player.cpp" and likewise for main i use lowercase as it was how the name in the codebase .
> 
> I think in that guide too , it is mentioned like that :
> 
> > > For C++ files, prefer naming the file the same as the (main) class it contains.

Yes, they mean you should / it is preferable to use lowercase for the file name even though the class name is capitalised. I think you misunderstood.

gounthar commented:
> @gounthar Did you suggest to have two branches for the code repo, with one for the skeleton and the other for the solution? If so, what is the motivation for doing so? I do not see why we need to ask the user to solve the problem as a challenge themselves, as that seems out of place. That is not what the tutorial is for to me.

I agree that this might be out of the scope of this repository.

biru-codeastromer commented:
> Yes, they mean you should / it is preferable to use lowercase for the file name even though the class name is capitalised and has the same sequence of characters. I think you misunderstood.


I think I got confused between the two lines . Thanks for your clarification .

biru-codeastromer commented:
Now my tasks are : 

1. Make the file name lowercase .
2. Keep one main branch as the ultimate branch . 
3. Use float for precision instead of int .


Should I move ahead now implementing them .🤗

krisstern commented:
Yes, please

biru-codeastromer commented:
Hello , I have made the changes as suggested . Needed to make few more changes for the new tests to works as expected and refined the needed files . 

Let me know you see the repo now after review . If it's good  : [Casino game repo](https://github.com/biru-codeastromer/casino-number-guessing-game)

biru-codeastromer commented:
I tested it in local : 

```
birajitsaikia@Birajits-MacBook-Pro casino-number-guessing-game % rm -rf build
birajitsaikia@Birajits-MacBook-Pro casino-number-guessing-game % cmake -B build -S .
-- The C compiler identification is AppleClang 16.0.0.16000026
-- The CXX compiler identification is AppleClang 16.0.0.16000026
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/birajitsaikia/Documents/c++ tutorial/casino-number-guessing-game/build
birajitsaikia@Birajits-MacBook-Pro casino-number-guessing-game % cmake --build build
[ 12%] Building CXX object CMakeFiles/casino_game.dir/src/main.cpp.o
[ 25%] Building CXX object CMakeFiles/casino_game.dir/src/Game.cpp.o
[ 37%] Building CXX object CMakeFiles/casino_game.dir/src/Player.cpp.o
[ 50%] Linking CXX executable casino_game
[ 50%] Built target casino_game
[ 62%] Building CXX object CMakeFiles/test_game.dir/tests/test_game.cpp.o
[ 75%] Building CXX object CMakeFiles/test_game.dir/src/Game.cpp.o
[ 87%] Building CXX object CMakeFiles/test_game.dir/src/Player.cpp.o
[100%] Linking CXX executable test_game
[100%] Built target test_game
birajitsaikia@Birajits-MacBook-Pro casino-number-guessing-game % ./build/casino_game
Welcome to the Casino Number Guessing Game!
Enter your name: birajit
Choose difficulty (1: Easy, 2: Medium, 3: Hard): 1
Generated number: 10
Enter your guess: 2
Wrong guess! Try again.
Enter your guess: 3
Wrong guess! Try again.
Enter your guess: 1
Wrong guess! Try again.
Enter your guess: 7
Wrong guess! Try again.
Enter your guess: 8
Wrong guess! Try again.
Enter your guess: 9
Wrong guess! Try again.
Enter your guess: 5
Wrong guess! Try again.
Enter your guess: 4
Wrong guess! Try again.
Enter your guess: 3
Wrong guess! Try again.
Enter your guess: 6
Wrong guess! Try again.
Enter your guess: 7
Wrong guess! Try again.
Enter your guess: 1
Wrong guess! Try again.
Enter your guess: 2
Wrong guess! Try again.
Enter your guess: 3
Wrong guess! Try again.
Enter your guess: 4
Wrong guess! Try again.
Enter your guess: 5
Wrong guess! Try again.
Enter your guess: 6
Wrong guess! Try again.
Enter your guess: 7
Wrong guess! Try again.
Enter your guess: 8
Wrong guess! Try again.
Enter your guess: 9
Wrong guess! Try again.
Enter your guess: 10
Congratulations! You guessed the number!
You won $4.76!
birajitsaikia@Birajits-MacBook-Pro casino-number-guessing-game % ./build/test_game
All tests passed!
```

krisstern commented:
@biru-codeastromer you have not update all file names in https://github.com/biru-codeastromer/casino-number-guessing-game/tree/main/src to lowercase.

krisstern commented:
Also regarding [L65 in "main.cpp"](https://github.com/biru-codeastromer/casino-number-guessing-game/blob/64c8b43de6056745e827d039af318da2ac380fc2/src/main.cpp#L65), where it says:
```c++
std::cout << "Congratulations! You guessed the number!" << std::endl;
```
Can we change it to
```c++
std::cout << "Congratulations! You guessed the number" << std::endl;
```
Otherwise it seems a tad bit too enthusiastic / unprofessional.

biru-codeastromer commented:
> @biru-codeastromer you have not update all file names in https://github.com/biru-codeastromer/casino-number-guessing-game/tree/main/src to lowercase.



I have changed in local , I guess something went wrong in the push.. It might be due to macOS case insensitive file system . I will fix that now .

biru-codeastromer commented:
I have made the changes for the case I did directly in github to make sure it changes.  : https://github.com/biru-codeastromer/casino-number-guessing-game/tree/main/src 

I applied your suggestion too . Thanks !

krisstern commented:
Looks good (enough) to me now! Thanks

krisstern commented:
TBH I already did indicated that when I wrote:

> Looks good (enough) to me now! Thanks

biru-codeastromer commented:
Sorry for that . I will do it now .

biru-codeastromer commented:
I raised the issue for that in the infra desk : https://github.com/jenkins-infra/jenkins-infra/issues/3917

biru-codeastromer commented:
Hello @gounthar 😇 , I got a reply from Damien Sir about the transfer as : https://github.com/jenkins-infra/helpdesk/issues/4572#issuecomment-2700932613

> @biru-codeastromer I'm closing this issue as "not planned" since the original repository is owned by you + the Jenkins Infra team has nothing to do with the jenkins-docs GitHub organization. Please check who are the right person with the Docs SIG

biru-codeastromer commented:
What should I do now ? Should I transfer the ownership in my repo to jenkins-doc org at first ? Please let me know how to proceed with it for the best way to transfer as I am not much familiar with it . It will be very much helpful . Thanks !

gounthar commented:
@kmartens27 , what are your thoughts?

krisstern commented:
@gounthar I can transfer the code to `jenkins-infra` but there are other things on my plate right now so will do so afterwards

gounthar commented:
@biru-codeastromer : could you try [that](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#transferring-a-repository-owned-by-your-personal-account)?

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2025-03-05 at 7 48 50 PM" src="https://github.com/user-attachments/assets/0525f2be-dbae-4ad9-9321-01a9b19a2c45" />


I get this when i go to settings and try to transfer @gounthar

gounthar commented:
I was kind of expecting that.
Let's go the fork route, then. 🤷

kmartens27 commented:
Sorry, just catching up on this one. It does look like it should be in the jenkins-docs org if anything. I'm only a member, and not an owner/admin of that so I don't think I can do any sort of transfer.

biru-codeastromer commented:
> I can't migrate your repo, as I am not an admin of it. 🤷

I will give that now to you

biru-codeastromer commented:
also I want to know is mirroring a better approach then forking ?

biru-codeastromer commented:
also I think if you become the admin of the repo now You can transfer the repo now to docs org ? am I right?

kmartens27 commented:
@biru-codeastromer mirroring creates a synchronized but read-only version, while forking allows the user to contribute work/modify content and submit it to the upstream. So the main aspect is being able to update the fork and work independently, as opposed to relying on the upstream maintainer to keep everything up to date or perform any needed work/maintenance.

gounthar commented:
I can now push to your repo, but my permissions aren't high enough to access your repo's settings.

biru-codeastromer commented:
> @biru-codeastromer mirroring creates a synchronized but read-only version, while forking allows the user to contribute work/modify content and submit it to the upstream. So the main aspect is being able to update the fork and work independently, as opposed to relying on the upstream maintainer to keep everything up to date or perform any needed work/maintenance.


Thanks a lot for letting me know the differences between them .Much appreciated .😊

biru-codeastromer commented:
> I can now push to your repo, but my permissions aren't high enough to access your repo's settings.

 I will now see to give high access to do that .

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2025-03-05 at 8 06 04 PM" src="https://github.com/user-attachments/assets/fef50018-d884-4939-9c6d-1d5ed271c6ff" />


I think for that the repo should be deleted from your repos list ?

gounthar commented:
It should be transferred to jenkins-docs, not gounthar.

gounthar commented:
But we can try this way, yes.

gounthar commented:
My bad. Done now.

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2025-03-05 at 8 21 38 PM" src="https://github.com/user-attachments/assets/795bab63-4a87-47aa-b559-59cfb5d24556" />

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2025-03-05 at 8 24 43 PM" src="https://github.com/user-attachments/assets/ec720f7b-c322-494c-a4b2-ff8b3bdc624d" />


I can see it is done  now !

gounthar commented:
We're good, thanks a lot for your help in that matter! 🙏

biru-codeastromer commented:
glad that we completed the task 🤗
